File: 0079-word-search/0079-word-search.java
class Solution {
     private char[][] matrix;
    private int ROWS;
    private int COLS;
    public boolean exist(char[][] board, String word) {
        this.matrix = board;
        this.ROWS = board.length;
        this.COLS = board[0].length;

        for(int row = 0; row < ROWS; row++){
            for(int col = 0; col < COLS; col++){
                if(backtrack(word,row,col,0)) 
                    return true;
            }
        }

        return false;
    }

    protected boolean backtrack(String word, int row, int col, int start){
        /* Step 1). check the bottom case. */
        if(start >= word.length())
            return true;

        /* Step 2). Check the boundaries. */
        if(row < 0 || col < 0 || row >= ROWS || col >= COLS || this.matrix[row][col] != word.charAt(start)){
            return false;
        }

        //Step 3) mark the path before the next exploration
        this.matrix[row][col] = '*';
        int[] rowOffsets = { 0, 1, 0, -1 };
        int[] colOffsets = { 1, 0, -1, 0 };

        /* Step 4). explore the neighbors in DFS */
        for(int dir = 0; dir < 4; dir++){
            if(backtrack(word, row + rowOffsets[dir], col + colOffsets[dir], start+1)){
                // return without cleanup
                return true;
            }
        }
        /* Step 5). clean up and return the result. */
        this.matrix[row][col] = word.charAt(start);
        return false;

    }

}

File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    int rows;
    int cols;

  bool backtrack(vector<vector<char>>&board , string word, int i , int j , int index)
  {
    if(index == word.size())
        return true;
    
    if(i >= rows or j >= cols or i < 0 or j < 0 or board[i][j] != word[index])
        return false;

    char t = board[i][j];
    board[i][j] = '*';

    int dx[4] = {1,-1,0,0};
    int dy[4] = {0,0,1,-1};
    
    bool x = backtrack(board,word,i+dx[0],j+dy[0],index+1) || 
            backtrack(board,word,i+dx[1],j+dy[1],index+1) ||
            backtrack(board,word,i+dx[2],j+dy[2],index+1) ||
            backtrack(board,word,i+dx[3],j+dy[3],index+1);

    board[i][j] = t;

    return x;
  }

  bool exist(vector<vector<char>>& board, string word)
  {
    rows = board.size();
    cols = board[0].size();

    for(int i=0;i<rows;i++)
    {
        for(int j=0;j<cols;j++)
        {
            if(backtrack(board,word,i,j,0))
                return true; 
        }
    }
    return false;
  }
 
};